druid/src/main/java/org/apache/calcite/adapter/druid/DruidDateTimeUtils.java

```diff
@@ -0,0 +1,529 @@
+    }
+    return null;
+  }
+
+  private static RexLiteral extractLiteral(RexNode node) {
+    RexNode target = node;
+    if (node.getKind() == SqlKind.CAST) {
+      target = ((RexCall) node).getOperands().get(0);
+    }
+    if (!(target instanceof RexLiteral)) {
+      return null;
+    }
+    return (RexLiteral) target;
+  }
+
+  private static Comparable toTimestamp(Object literal) {
+    if (literal instanceof Timestamp) {
+      return (Timestamp) literal;
+    }
+    if (literal instanceof Date) {
+      return new Timestamp(((Date) literal).getTime());
+    }
+    if (literal instanceof Calendar) {
+      return new Timestamp(((Calendar) literal).getTime().getTime());
+    }
+    if (literal instanceof Number) {
+      return new Timestamp(((Number) literal).longValue());
+    }
+    if (literal instanceof String) {
+      String string = (String) literal;
+      if (StringUtils.isNumeric(string)) {
+        return new Timestamp(Long.valueOf(string));
+      }
+      try {
+        return Timestamp.valueOf(string);
+      } catch (NumberFormatException e) {
+        // ignore
+      }
+    }
+    return null;
+  }
+
+  private static Long toLong(Object literal) {
+    if (literal instanceof Number) {
+      return ((Number) literal).longValue();
+    }
+    if (literal instanceof Date) {
+      return ((Date) literal).getTime();
+    }
+    if (literal instanceof Timestamp) {
+      return ((Timestamp) literal).getTime();
+    }
+    if (literal instanceof Calendar) {
+      return ((Calendar) literal).getTime().getTime();
+    }
+    if (literal instanceof String) {
+      try {
+        return Long.valueOf((String) literal);
+      } catch (NumberFormatException e) {
+        // ignore
+      }
+      try {
+        return DateFormat.getDateInstance().parse((String) literal).getTime();
+      } catch (ParseException e) {
+        // best effort. ignore
+      }
+    }
+    return null;
+  }
+
+
+  private static Integer toInt(Object literal) {
+    if (literal instanceof Number) {
+      return ((Number) literal).intValue();
+    }
+    if (literal instanceof String) {
+      try {
+        return Integer.valueOf((String) literal);
+      } catch (NumberFormatException e) {
+        // ignore
+      }
+    }
+    return null;
+  }
+
+  private static Float toFloat(Object literal) {
+    if (literal instanceof Number) {
+      return ((Number) literal).floatValue();
+    }
+    if (literal instanceof String) {
+      try {
+        return Float.valueOf((String) literal);
+      } catch (NumberFormatException e) {
+        // ignore
+      }
+    }
+    return null;
+  }
+
+  private static Double toDouble(Object literal) {
+    if (literal instanceof Number) {
+      return ((Number) literal).doubleValue();
+    }
+    if (literal instanceof String) {
+      try {
+        return Double.valueOf((String) literal);
+      } catch (NumberFormatException e) {
+        // ignore
+      }
+    }
+    return null;
+  }
+
```

Comment:
Indeed I have removed the condenseRanges logic... I should have figured out I could use RangeSet for that.